### PR TITLE
State migration

### PR DESCRIPTION
**Pull request by husio**
x/lateinit provides a way to declare and execute a one time state
modification by creating an entity. This extension is intended to be
used in a similar fashion to genessis, but for already initialized
chains.

@alpe @ethanfrey `weave` allows for a state migration! :dancer: 

Lateinit provides a way to create a defined in code entity in order to
change the current state of the chain. To apply such modification two
steps are required.
1. Declare a state migration in code and build new bnsd binary. Distribute the
   new binary to all validators.
2. Send an `lateinit.ExecuteInitMsg` with information which
   initialization to execute.

Once an initialization is declared in code and executed, it must never
change. This is crucial for the state rebuilding.

Each initialization can be executed only once. Subsequent execution
attempts will fail.

Example migration declaration:

https://github.com/iov-one/weave/blob/e669495a3bd6b4bbbeeebe91502c763988503aba/cmd/bnsd/app/lateinit.go#L12-L20

### CODE REVIEW

**Comment by husio** [comment]
As @webmaster128 mentioned, this functionality allows to define a [wallet creation](https://github.com/iov-one/weave/blob/4f6004889f9e9f755730a05b8466fb277fe28665/x/cash/codec.proto#L9-L14] and therefore materializing coins. I could for example declare a state migrations, creating owned by me wallet.

```go
lateinit.MustRegisterGConfig(                                                  
        "x/cash/piotr_is_getting_rich",                                        
        weave.Address("a_key_owned_by_piotr"),                                 
        &msgfee.Configuration{                                                 
                Metadata: &weave.Metadata{Schema: 1},                          
                Coins: []coin.Coin{                                            
                        Whole: 1000000,                                        
                        Ticker: "IOV",                                         
                }                                                              
        },                                                                     
})
```

If I will manage to merge and convince validators to deploy this code, I will materialize an account with [1 million IOV tokens](https://upload.wikimedia.org/wikipedia/en/1/16/Drevil_million_dollars.jpg).

This implementation allows only for entity creation. It does not allow for update or deletion.

cc: @davepuchyr

**Comment by webmaster128** [comment]
> this functionality allows to define a wallet creation and therefore materializing coins. I could for example declare a state migrations, creating owned by me wallet.

Great, thanks for checking and clarifying. I think it is important to develop and understanding of the potential and the limitations of this module. Glad we're on the same page.

File: cmd/bnsd/app/lateinit.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import (
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/x/lateinit"
+	"github.com/iov-one/weave/x/msgfee"
+)
+
+func init() {
+	// cond:gov/rule/0000000000000003 is owned by technical executors
+	lateinit.MustRegisterGConfig(
+		"msgfee",
```
Comment (webmaster128) [inline cmd/bnsd/app/lateinit.go]: Is this this init ID? Can we make it a little more self-explanary, like "set_msgfee_admin_to_economic_executors"

File: cmd/bnsd/app/lateinit.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+/*
+import (
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/x/lateinit"
+	"github.com/iov-one/weave/x/msgfee"
+)
+
+func init() {
+	// cond:gov/rule/0000000000000003 is owned by technical executors
```
Comment (webmaster128) [inline cmd/bnsd/app/lateinit.go]: If I understand correctly, the `cond:gov/rule/0000000000000002` (economic executors) should set the product fees and thus should be admin. @davepuchyr?

The owner however could be the technical executors

File: cmd/bnsd/app/lateinit.go (hunk 1 above)
Comment (orkunkl) [reply cmd/bnsd/app/lateinit.go]: That is the package name that the initialization will be run on.
```go
func MustRegisterGConfig(
	pkgName string,
	requiredSigner weave.Address,
	entity orm.Model,
) {
	MustRegister(
		pkgName+"/configuration",
		requiredSigner,
		[]byte(pkgName),
		gconf.NewConfigurationModelBucket(),
		entity,
	)
}
```
`MustRegister` merges `pkgName` and `/configuration` string, which becomes `msg_fee/configuration` in this instance. Then it is fed to MustRegister as `initID`:
```go
// MustRegister registers an initialization function for given entity.
func MustRegister(
	initID string,
	requiredSigner weave.Address,
	entityID []byte,
	bucket orm.ModelBucket,
	entity orm.Model,
)
```

**Comment by orkunkl** [comment]
> As @webmaster128 mentioned, this functionality allows to define a [wallet creation](https://github.com/iov-one/weave/blob/4f6004889f9e9f755730a05b8466fb277fe28665/x/cash/codec.proto#L9-L14) and therefore materializing coins. I could for example declare a state migrations, creating owned by me wallet.
> 
> ```go
> lateinit.MustRegisterGConfig(                                                  
>         "x/cash/piotr_is_getting_rich",                                        
>         weave.Address("a_key_owned_by_piotr"),                                 
>         &msgfee.Configuration{                                                 
>                 Metadata: &weave.Metadata{Schema: 1},                          
>                 Coins: []coin.Coin{                                            
>                         Whole: 1000000,                                        
>                         Ticker: "IOV",                                         
>                 }                                                              
>         },                                                                     
> })
> ```
> 
> If I will manage to merge and convince validators to deploy this code, I will materialize an account with [1 million IOV tokens](https://upload.wikimedia.org/wikipedia/en/1/16/Drevil_million_dollars.jpg).
> 
> This implementation allows only for entity creation. It does not allow for update or deletion.
> 
> cc: @davepuchyr

We should let validators be aware of this situation before the merge

**Comment by husio** [comment]
> As soon as some validators install the new software containing a new initialization migration, sending the lateinit.ExecuteInitMsg leads to different transaction execution results, which can break consensus. To mitigate this problem, I think we need a minHeight or minTime at which a migration becomes valid. This can be set to a time point in the future at which a majority of the validators is expeced to be upgraded.

I was thinking about this and I think time/height assurance is not needed. This is because a migration can only be executed by a single address. Admin should not wish harm to a network.

**Comment by husio** [comment]
While this implementation solves the problem of configuration initialization, it does not solve a real data migration needs.

We are about to implement an index for accounts https://github.com/iov-one/weave/issues/1055
One of the acceptance criteria points is that all existing accounts are indexed as well. To fulfill this requirement, we must run a code that will determine the data migration at runtime. We want to be able to execute below pesudocode:
```
def migrate_index_accounts(db):
    for red in list_red_accounts(db):
        index(db, red)
    for blue in list_blue_accounts(db):
        index(db, blue)
```

The current implementation is nice, but it is not enough. My proposal is to instead of registering an entity to be inserted, register a function that will be executed. This is much more powerful mechanism and allows introducing bugs if used badly. It is also necessary for account index migration. 

What do you guys think?

**Comment by davepuchyr** [comment]
> My proposal is to instead of registering an entity to be inserted, register a function that will be executed. This is much more powerful mechanism and allows introducing bugs if used badly. It is also necessary for account index migration.
> 
> What do you guys think?

I think that function registration is the way to go.  If someone wants to shoot themselves in the foot by writing a faulty function then so be it.  An SDK can't guard against business logic bugs.

**Comment by husio** [comment]
We need a more powerful solution.

**Pull request closed without merging**